src/controllers/resource-utils/index.ts

```diff
@@ -1,39 +1,39 @@
-import { RouteResource, RouteResourceGettersArgs } from '../../common/types';
+import {RouteResource, RouteResourceDataPayload, RouteResourceGettersArgs} from '../../common/types';
 import { DEFAULT_RESOURCE_MAX_AGE } from '../resource-store/constants';
 
 /**
  * Utility method to created async versions of getData functions
  *
  */
-type GetDataLoader = () => Promise<{
-  default: RouteResource['getData'];
+type GetDataLoader<T> = () => Promise<{
+  default: RouteResource<T>['getData'];
 }>;
 
 type BaseResource = Pick<RouteResource, 'type' | 'getKey'>;
 
-interface CreateResourceSync extends BaseResource {
-  getData: RouteResource['getData'];
+interface CreateResourceSync<T> extends BaseResource {
+  getData: RouteResource<T>['getData'];
   maxAge?: number;
 }
-interface CreateResourceAsync extends BaseResource {
+interface CreateResourceAsync<T> extends BaseResource {
   getDataLoader: (
     ...args: RouteResourceGettersArgs
   ) => Promise<{
-    default: GetDataLoader;
+    default: GetDataLoader<T>;
   }>;
   maxAge?: number;
 }
 
-const handleGetDataLoader = (asyncImport: GetDataLoader) => {
+const handleGetDataLoader = (asyncImport: GetDataLoader<unknown>) => {
   return async (...args: RouteResourceGettersArgs) => {
     const { default: getDataFn } = await asyncImport();
 
     return getDataFn(...args);
   };
 };
 
-export function createResource(args: CreateResourceSync): RouteResource;
-export function createResource(args: CreateResourceAsync): RouteResource;
+export function createResource<T extends unknown = RouteResourceDataPayload>(args: CreateResourceSync<T>): RouteResource<T>;
```

Comment:
note - by default `T` is equal to the previously used shape (Record) to prevent any possible breaking change.